src/System.Buffers.Primitives/System/Internal/ManagedBufferPool.cs

```diff
@@ -89,7 +90,10 @@ public override void Retain()
             public override void Release()
             {
                 if (!IsRetained) BufferPrimitivesThrowHelper.ThrowInvalidOperationException();
-                Interlocked.Decrement(ref _referenceCount);
+                if(Interlocked.Decrement(ref _referenceCount) <= 0)
```

Comment:
Yeah, you are right. I was trying to avoid some checks, but in the absence of synchronization around IsRetianed, the optimizations won't work. I will fix.